src/solvers/dgsem_tree/indicators_3d.jl

```diff
@@ -135,6 +135,224 @@ function (indicator_hg::IndicatorHennemannGassner)(u::AbstractArray{<:Any,5},
+    for k in 1:nnodes(dg), j in 1:nnodes(dg), i in 1:nnodes(dg)
+      total_energy += modal[i, j, k]^2
+    end
+    total_energy_clip1 = zero(eltype(modal))
+    for k in 1:(nnodes(dg)-1), j in 1:(nnodes(dg)-1), i in 1:(nnodes(dg)-1)
+      total_energy_clip1 += modal[i, j, k]^2
+    end
+    total_energy_clip2 = zero(eltype(modal))
+    for k in 1:(nnodes(dg)-2), j in 1:(nnodes(dg)-2), i in 1:(nnodes(dg)-2)
+      total_energy_clip2 += modal[i, j, k]^2
+    end
+
+    # Calculate energy in lower modes
+    energy = max((total_energy - total_energy_clip1) / total_energy,
+                 (total_energy_clip1 - total_energy_clip2) / total_energy_clip1)
+
+    alpha_element = 1 / (1 + exp(-parameter_s / threshold * (energy - threshold)))
+
+    # Take care of the case close to pure DG
+    if alpha_element < alpha_min
+      alpha_element = zero(alpha_element)
+    end
+
+    # Take care of the case close to pure FV
+    if alpha_element > 1 - alpha_min
+      alpha_element = one(alpha_element)
+    end
+
+    # Clip the maximum amount of FV allowed
+    alpha[element] = min(alpha_max, alpha_element)
+  end
+
+  if (alpha_smooth)
+    # Diffuse alpha values by setting each alpha to at least 50% of neighboring elements' alpha
+    # Copy alpha values such that smoothing is indpedenent of the element access order
+    alpha_tmp .= alpha
+
+    # Loop over elements, because there is no interface container
+    for element in eachelement(dg,cache)
+      # Get neighboring element ids
+      left  = cache.elements.left_neighbors[1, element]
+      right = cache.elements.right_neighbors[1, element]
+      up    = cache.elements.left_neighbors[2, element] 
+      down  = cache.elements.right_neighbors[2, element]
+      front = cache.elements.left_neighbors[3, element]
+      back  = cache.elements.right_neighbors[3, element]
+
+      # Apply smoothing
+      alpha[left]     = max(alpha_tmp[left],    0.5 * alpha_tmp[element], alpha[left])
+      alpha[element]  = max(alpha_tmp[element], 0.5 * alpha_tmp[left],    alpha[element])
+      
+      alpha[right]    = max(alpha_tmp[right],   0.5 * alpha_tmp[element], alpha[right])
+      alpha[element]  = max(alpha_tmp[element], 0.5 * alpha_tmp[right],   alpha[element])
+
+      alpha[up]       = max(alpha_tmp[up],      0.5 * alpha_tmp[element], alpha[up])
+      alpha[element]  = max(alpha_tmp[element], 0.5 * alpha_tmp[up],      alpha[element])
+
+      alpha[down]     = max(alpha_tmp[down],    0.5 * alpha_tmp[element], alpha[down])
+      alpha[element]  = max(alpha_tmp[element], 0.5 * alpha_tmp[down],    alpha[element])
+      
+      alpha[front]    = max(alpha_tmp[front],   0.5 * alpha_tmp[element], alpha[front])
+      alpha[element]  = max(alpha_tmp[element], 0.5 * alpha_tmp[front],   alpha[element])
+
+      alpha[back]     = max(alpha_tmp[back],    0.5 * alpha_tmp[element], alpha[back])
+      alpha[element]  = max(alpha_tmp[element], 0.5 * alpha_tmp[back],    alpha[element])
+
+    end
+
+  end
```

Comment:
It looks to me like only this smoothing part needs to be implemented differently. Is smoothing specific to this indicator, or is it done with different indicators as well? Would it make sense to extract this smoothing function, so that indicators can be generic without bothering with mesh-specific code, and then the smoothing is a mesh-specific "core" function of Trixi (in the sense that indicators are generic code that can be written quickly in elixirs, while the smoothing is one of those functions with many different implementations that no one will ever "quickly" write in an elixir)?
This is a question to @sloede and @ranocha.

Reply:
The smoothing should definitely be moved to another function (which can be specialized on the `mesh`)

Reply:
done in 3bad0d4.